### ISSUE TITLE
doesn't remove unicode BOM from empty file

### ISSUE THREAD

**Issue by evilebottnawi** [prettier/prettier#6461]
**Prettier 1.18.2**
No playground.

For example using `CLI`:
```
prettier --write test.js
```

**Input:**

**Empty file with unicode BOM**

```jsx
```

**Output:**

**Empty file with unicode BOM**

```jsx
```

**Expected behavior:**

**Empty file without unicode BOM**

```jsx
```

If you add content BOM will be remove.

**.editorconfig**
```
# For more information please visit http://editorconfig.org
root = true

[*]
indent_style = space
indent_size = 2
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true
```

**Comment by evilebottnawi** [comment]
@lydell i found we recommend disable `unicode-bom` rule https://github.com/prettier/eslint-config-prettier/blob/master/index.js#L89, but i don't find code that removing or adding BOM. 

We respect BOM from source (i.e. if BOM exists we return this, but if no BOM we don't add this) https://github.com/prettier/prettier/blob/master/src/main/core.js#L356. 

Why we recommend remove disable rule?

Also i think we can improve this - auto removing/adding BOM based on `editorconfig`. 

For example if you have
```
charset = utf-8
```

we remove BOM

If you have 
```
charset = utf-8-bom
```

we add BOM,

it is not hard

**Comment by lydell** [comment]
`"unicode-bom": "off",` has been present in eslint-config-prettier since the initial commit. I think it’s just an oversight, or Prettier used to deal with BOM differently in January 2017, I can’t remember.

**Comment by evilebottnawi** [comment]
@lydell i think we should remove this rule right now, after we implement logic above

**Comment by fisker** [comment]
@evilebottnawi Only if user has `.editorconfig` right? Otherwise we still respect from source.

**Comment by fisker** [comment]
I don't think it's bug.

**Comment by evilebottnawi** [comment]
@fisker it is very weird, because `eslint-config-prettier` recommend to disable `unicode-bom` rule, but we don't touch BOM, so we should not recommend to disable it

**Comment by lydell** [comment]
Yeah, if Prettier preserves BOM/no-BOM of files then it is a bug in eslint-config-prettier to disable that rule

**Comment by Maxim-Mazurok** [comment]
> if you have
> 
> ```
> charset = utf-8
> ```
> 
> we remove BOM
> 
> If you have
> 
> ```
> charset = utf-8-bom
> ```
> 
> we add BOM

For some reason, I assumed that prettier already does that, and was surprised to find out that it doesn't...

It keeps EOL consistent with `.editorconfig`, so why wouldn't it do the same for charset?

**Comment by lydell** [comment]
FYI, [eslint-config-prettier 9.0.0](https://github.com/prettier/eslint-config-prettier/blob/19826807f2d668a05bb9c29a5f6f6a6e6e3287e4/CHANGELOG.md#version-900-2023-08-05) finally stops disabling the `unicode-bom` ESLint rule.
